agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpClientTracer.java

```diff
@@ -0,0 +1,179 @@
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.TracingContextUtils;
+import io.opentelemetry.trace.attributes.SemanticAttributes;
+import java.net.URI;
+import java.net.URISyntaxException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class HttpClientTracer<REQUEST, RESPONSE> extends BaseTracer {
+
+  private static final Logger log = LoggerFactory.getLogger(HttpClientTracer.class);
+
+  public static final String DEFAULT_SPAN_NAME = "HTTP request";
+
+  protected static final String USER_AGENT = "User-Agent";
+
+  protected abstract String method(REQUEST request);
+
+  protected abstract URI url(REQUEST request) throws URISyntaxException;
+
+  protected abstract Integer status(RESPONSE response);
+
+  protected abstract String requestHeader(REQUEST request, String name);
+
+  protected abstract String responseHeader(RESPONSE response, String name);
+
+  protected HttpTextFormat.Setter<REQUEST> getSetter() {
+    throw new UnsupportedOperationException();
+  }
+
+  public Span startSpan(REQUEST request) {
+    return startSpan(request, spanNameForRequest(request));
+  }
+
+  public Scope startScope(Span span, REQUEST request) {
+    Context context = withSpan(span, Context.current());
+    OpenTelemetry.getPropagators().getHttpTextFormat().inject(context, request, getSetter());
+    return withScopedContext(context);
+  }
+
+  // TODO should end methods remove SPAN attribute from request as well?
+  public void end(Span span, RESPONSE response) {
+    onResponse(span, response);
+    super.end(span);
+  }
+
+  public void endExceptionally(Span span, RESPONSE response, Throwable throwable) {
+    onResponse(span, response);
+    super.endExceptionally(span, throwable);
+  }
+
+  /**
+   * Returns a new client {@link Span} if there is no client {@link Span} in the current {@link
+   * Context}, or an invalid {@link Span} otherwise.
+   */
+  private Span startSpan(REQUEST request, String name) {
+    Context context = Context.current();
+    Span clientSpan = ClientDecorator.CONTEXT_CLIENT_SPAN_KEY.get(context);
+
+    if (clientSpan != null) {
+      // We don't want to create two client spans for a given client call, suppress inner spans.
+      return DefaultSpan.getInvalid();
+    }
+
+    Span current = TracingContextUtils.getSpan(context);
+    Span span = tracer.spanBuilder(name).setSpanKind(Kind.CLIENT).setParent(current).startSpan();
+    onRequest(span, request);
+    return span;
+  }
+
+  private Span onRequest(final Span span, final REQUEST request) {
+    assert span != null;
+    if (request != null) {
+      span.setAttribute(SemanticAttributes.HTTP_METHOD.key(), method(request));
+
+      String userAgent = requestHeader(request, USER_AGENT);
+      if (userAgent != null) {
+        SemanticAttributes.HTTP_USER_AGENT.set(span, userAgent);
+      }
+
+      // Copy of HttpServerDecorator url handling
```

Comment:
Can we extract a helper?

Reply:
And this is duplicated with `HttpServerTracer`. Certainly have to de-duplicate.

Reply:
Will clean it up later with de-duplicate client and server code.  I have noted it down as a reminder for myself. Can someone create a new issue for cleaning up common code between client and server tracers and assign it to me.  Thanks.